onnxruntime/core/util/qmath.h

```diff
@@ -305,4 +305,591 @@
+   *
+   * @param thread_pool           thread pool
+   * @param input                 input tensor
+   * @param scale                 scale tensor
+   * @param zero_point            zero point tensor
+   * @param output                output tensor
+   * @param M                     total size of dimensions before quantize axis
+   * @param K                     size of dimension on quantize axis
+   * @param N                     total size of dimensions after quantize axis
+   * @param quant_block_size      quantization block size
+   * @param block_size            task block size
+   * @param saturate              used by float8
+   */
+  static void opNotLastAxis(concurrency::ThreadPool* thread_pool, const float* input, const float* scale,
+                            const TOut* zero_point, TOut* output, std::ptrdiff_t M, std::ptrdiff_t K,
+                            std::ptrdiff_t N, const std::ptrdiff_t quant_block_size,
+                            const std::ptrdiff_t block_size, bool saturate);
+
+  /**
+   * @brief Compute blocked quantization using multi-threading and vector instructions.
+   *        Quantize axis is the last axis. Block along quantize axis using quant_block_size
+   *        as block_size. quant_block_size is usually 2's power between 16 and 256.
+   *        Within a block, scale index does not change.
+   *
+   * @param thread_pool           thread pool
+   * @param input                 input tensor
+   * @param scale                 scale tensor
+   * @param zero_point            zero point tensor
+   * @param output                output tensor
+   * @param M                     total size of dimensions before quantize axis
+   * @param K                     size of dimension on quantize axis
+   * @param quant_block_size      quantization block size
+   * @param saturate              used by float8
+   */
+  static void opLastAxis(concurrency::ThreadPool* thread_pool, const float* input, const float* scale,
+                         const TOut* zero_point, TOut* output, std::ptrdiff_t M, std::ptrdiff_t K,
+                         const std::ptrdiff_t quant_block_size, bool saturate);
+};
+
+template <typename TOut>
+struct BlockedQuantizeLinear<float, TOut, 0> {
+  static void opNotLastAxis(concurrency::ThreadPool* thread_pool, const float* input, const float* scale,
+                            const TOut* zero_point, TOut* output, std::ptrdiff_t M, std::ptrdiff_t K,
+                            std::ptrdiff_t N, const std::ptrdiff_t quant_block_size,
+                            const std::ptrdiff_t block_size, bool saturate) {
+    ORT_UNUSED_PARAMETER(saturate);
+    constexpr auto low = static_cast<int32_t>(std::numeric_limits<TOut>::lowest());
+    constexpr auto high = static_cast<int32_t>(std::numeric_limits<TOut>::max());
+    const auto num_r_block = (N + block_size - 1) / block_size;
+    const auto num_blocks = M * K * num_r_block;
```

Comment:
>M * K [](http://example.com/codeflow?start=28&length=5)

in real scenario, M * K is large usually. Maybe you don't need a block_size.

Reply:
this is to calculate number of thread blocks, so the number depends on the thread block size.